trunk/NDHMS/Routing/Reservoirs/Persistence_Level_Pool_Hybrid/module_persistence_levelpool_hybrid.F

```diff
@@ -0,0 +1,371 @@
+        type (hybrid_parameters_interface), pointer :: parameters => null()
+        type (hybrid_state_interface), pointer :: state => null()
+
+        logical :: pointer_allocation_guard = .false.
+
+    contains
+
+        procedure :: init => hybrid_type_init
+        procedure :: destroy => hybrid_type_destroy
+        procedure :: run => run_hybrid_reservoir
+
+    end type persistence_levelpool_hybrid_type
+
+#ifndef NCEP_WCOSS
+    integer, parameter :: log_warning = 6
+#else
+    integer, parameter :: log_warning = 78
+#endif
+
+contains
+
+    ! Hybrid Constructor
+    subroutine hybrid_type_init(this, water_elevation,  &
+        lake_area, weir_elevation, weir_coeffecient, &
+        weir_length, orifice_elevation, orifice_coefficient, &
+        orifice_area, lake_max_water_elevation, initial_fractional_depth, &
+        lake_number, persistence_parameter_file, start_date, &
+        timeslice_path, observation_lookback_hours, &
+        observation_update_time_interval_seconds)
+        implicit none
+        class(persistence_levelpool_hybrid_type), intent(inout) :: this ! object being initialized
+        real,    intent(inout) :: water_elevation           ! meters AMSL
+        real,    intent(in)    :: lake_area      		    ! area of lake (km^2)
+        real,    intent(in)    :: weir_elevation            ! bottom of weir elevation (meters AMSL)
+        real,    intent(in)    :: weir_coeffecient          ! weir coefficient
+        real,    intent(in)    :: weir_length               ! weir length (meters)
+        real,    intent(in)    :: orifice_elevation         ! orifice elevation (meters AMSL)
+        real,    intent(in)    :: orifice_coefficient       ! orifice coefficient
+        real,    intent(in)    :: orifice_area              ! orifice area (meters^2)
+        real,    intent(in)    :: lake_max_water_elevation  ! max water elevation (meters)
+        real,    intent(in)    :: initial_fractional_depth  ! initial fraction water depth
+        integer, intent(in)    :: lake_number               ! lake number
+        character(len=*),   intent(in) :: persistence_parameter_file
+        character(len=19),  intent(in) :: start_date
+        character(len=256), intent(in) :: timeslice_path
+        integer,            intent(in) :: observation_lookback_hours
+        integer,            intent(in) :: observation_update_time_interval_seconds
+        character(len=15)              :: lake_number_string
+
+#ifdef RESERVOIR_D
+        ! Create diagnostic log file only for development/debugging purposes
+        call create_hybrid_diagnostic_log_file(lake_number)
+#endif
+
+        ! Call to initialize read timeslice data object. This object is a singleton,
+        ! so if another reservoir has already initialized the object, it will return
+        ! immediately.
+        call timeslice_data%init(start_date, &
+            timeslice_path, persistence_parameter_file, observation_lookback_hours)
```

Comment:
I'm concerned about file contention on conus when there are >500 processors trying to read the same file. The only way to know if this is not going to be a problem is to try it out. Until then I will expect a problem. Can we test this on WCOSS? Or even on cheyenne. Run the code with and with out the timeslices and see the difference.

Reply:
What do you consider to be "file contention"?

Let me try to address this in a little more detail.  There are 240 USGS persistence reservoirs, and these are distributed across the domain and more than one may end up in the same process's sub-domain.  In terms of forecasting, these files are read ONE time per process that contains one (or more) persistence reservoir.  When multiple processes on a NODE are READING the same file, they each can read these independently with no contention (if opened in read mode, the OS can even help optimize these reads.) There could potentially be some small additional latency if the OS does try to read the file multiple times from the underlying disk, but these should be minimal given reasonable hardware caches and pre-fetching common in most modern architectures.

For reference, here are some of the times from the evaluation testing done on Cheyenne for forecasts.  These are on the higher end of the times based on Mehdi's comments:
    10 day re-forecast
     Levelpool Only: resources_used.walltime=00:57:09
     Persistence: resources_used.walltime=00:47:48

The AnA may suffer a little more from I/O issues simply because it reads the obs more often (right now, every hour of simulation time).  Again only for the first reservoir to call update per processes, and recall not every process has reservoirs.  Here are some sample breakdowns from the evaluation testing on Cheyenne.  Remember these are CHANNEL ONLY configurations. The AnA was run in 24 day chunks, I've normalized this to SimHours per WallHour:
    Levelpool Only: 164.6 SimHours/WallHour
    Persistence: 104.7 SimHours/WallHour

These are rough estimates from a sample of runs.  I do think it would be a good exercise to set up an actual timing test that is controlled enough and sampled correctly to provide significance.  The forecasts really should not have an impact (and may even be faster???) . The AnA could suffer, but there is a mechanism to alleviate the I/O burden at a small sacrifice to observational integration frequency. (Reading updates every 4 hours instead of 1 would reduce that burden significantly.)

Reply:
Thanks for detailing that. I'm still worried about relying "reasonable hardware caches and pre-fetching common in most modern architectures" but I will accept this argument with the data provided and I can remain a worrywart.

The run time is 20% greater, seems considerable. I'm unable to judge if that's proportional to the data. 
Alas, we have planned code performance/timing tests for ages, so that performance is benchmarked into a database with each PR..... 

I will just say that I would not be surprised if this issue comes up on WCOSS, but that does not mean it will. But then that bridge can be crossed.

Reply:
Wait. I was reading those numbers backwards. I'm going to say that additional computation and IO speeding things up is a cheyenne noise problem.